### ISSUE TITLE
[java] TooFewBranchesForSwitch false positive for exhaustive switches over enums without default case

### ISSUE THREAD

**Issue by esc-sbarden** [pmd/pmd#5311]
**Affects PMD Version:** 7.7.0

**Rule:** [TooFewBranchesForSwitch](https://docs.pmd-code.org/pmd-doc-7.7.0/pmd_rules_java_performance.html#toofewbranchesforswitch)

**Description:**
When Switch with enums, this rule reports a violations when there are too few branches.

Using switch is preferable, as the compiler will generate a compiler error if a new enum value is added and the switch doesn't cover all cases. Would the switch be rewritten with if-else then there will never be a compiler error if a new enum value appears.

**Code Sample demonstrating the issue:**

```java
public class Foo {

    String foo(Bar bar) {
        return switch (bar) {
            case A -> "A";
            case B -> "B";
        };
    }

    enum Bar {
        A,
        B
    }
}
```

**Expected outcome:**

PMD reports a violation at line ..., but that's wrong. That's a false positive.
Even thoug this switch has only two cases this is preferable, as when a new Enum value `C` is added, the compiler will generate an error.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

**Comment by julius-d** [comment]
Hi! 👋 

I also just run into that issue.
I wanted to update pmd from 7.6.0 to 7.7.0 
and suddenly got multiple `TooFewBranchesForASwitchStatement` warnings
for code like:

```java
public class Foo {

  String foo(Bar bar) {
    return switch (bar) {
      case A, B, C, D -> "Hello";
      case E, F, G -> "World";
    };
  }

  enum Bar {
    A, B, C, D, E, F, G
  }
}
```


and I am sure that replacing that `switch` with an `if` is bad in multiple ways:
```java
public class Foo {

  String foo(Bar bar) {
      if (bar == A || bar == B || bar == C || bar == D) {
          return "Hello";
      } else if (bar == E || bar == F || bar == G) {
          return "World";
      }
      throw new IllegalArgumentException("Unknown bar "+bar);
  }

  static enum Bar {
    A, B, C, D, E, F, G
  }
}
```

**Comment by adangel** [comment]
Duplicate of #5287

**Comment by adangel** [comment]
Ok, the comment https://github.com/pmd/pmd/issues/5311#issuecomment-2462867396 mislead me - the comment is a duplicate of #5287. But the comment is a different case, then this issue.
The comment is about a list of case constants, but this issue is just about switch over enum with few branches _without a list of case constants_.

**Comment by adangel** [comment]
@esc-sbarden I could imagine, that we adjust the rule to not report a violation under the following condition: the switch expression is a enum type and the switch doesn't have a default case.

I assume, this is not a new behavior of this rule - it always reported such cases, and this is actually the example case on https://docs.pmd-code.org/pmd-doc-7.7.0/pmd_rules_java_performance.html#toofewbranchesforswitch - the example there has one case and a default case.

So, from that perspective - the rule works as designed.

However, I already excluded switches, that use patterns - as for these, the compiler will do the exhaustiveness check - and that's what we want. That's the same argument you mention ("compiler will generate a compiler error if a new enum value is added and the switch doesn't cover all cases.").
Note, that we actually need a new rule, that find unnecessary default cases for switches, that already cover all cases then (ref #5246)

**Comment by Bukama** [comment]
Good morning,
I'm just updating our rule set due the changes with 7.7.0 and this rule is also listed in the deprecated warnings, so I had a look at it.

My thoughts (I hope I understand the changes in your PR correctly): Remove the rule completly. With pattern matching you also want to use switch to ensure to cover all cases, regardless if you have two classes which exhaust the switch or two million. See #4813 which shows the need to adjust SwitchStmtsShouldHaveDefault when using pattern matching.

**Comment by adangel** [comment]
> Remove the rule completly

No, it has been renamed: https://docs.pmd-code.org/latest/pmd_rules_java_performance.html#toofewbranchesforaswitchstatement :

> This rule has been renamed. Use instead: [TooFewBranchesForSwitch](https://docs.pmd-code.org/latest/pmd_rules_java_performance.html#toofewbranchesforswitch)

Notice, the only difference is, that this rule now also correctly considers switch expressions and not only statements.

With pattern matching, it's a bit different (and similar to enums): For those, the compiler can check the exhaustiveness for you and your code won't compile, if you don't cover all cases. That means, a switch statement/expression, that has only one case (pattern matching), and compiles - should be ok. Currently the rule still detects this case - and that's the topic of this issue here.

Note: As soon as you add a default case, you bypass your compiler's check for exhaustiveness and this means, a switch statement/expression with only one case (pattern matching) and a default case should still be reported, as there you don't gain any benefit by using a switch over a simple if-stmt and actually create a risk for future changes of your code (you won't get a compiler error, if the switch statement is not exhaustive anymore).

**Comment by Bukama** [comment]
> That means, a switch statement/expression, that has only one case (pattern matching), and compiles - should be ok. Currently the rule still detects this case - and that's the topic of this issue here.

Yeah exactly. And my _personal_ point of view to this is that the usage of this rule became obsolete with pattern matching, where you have valid situations (secured by the compiler) with less than 3 branches, where this rule would (and is) raise a warning. 

> Note: As soon as you add a default case, you bypass your compiler's check for exhaustiveness and this means, a switch statement/expression with only one case (pattern matching) and a default case should still be reported, as there you don't gain any benefit by using a switch over a simple if-stmt and actually create a risk for future changes of your code (you won't get a compiler error, if the switch statement is not exhaustive anymore).

This is true (and why the JEP advised to not use a default branch at all), but I doubt this rule as of now has a different intention. Your scenario you've set up doesn't really fit the problem of the "dangerous" default branch. If you change your scenario to three branches + default branch but five possible outcomes of the pattern matching the rule would never apply (assuming default value of 3), but a rules would be needed which tasks it is to detect "stupid default branches".

Don't get me wrong I think we are not far from each other, but I think with pattern matching the aim of this rule is not as important as in the past (as there are other rules to cover other aspects of the switch statement, e.g. no default branch - which also needs to work with pattern matching (and NonExhaustiveSwitch does so).
